README.md
<div align="center">
  <img src="vignettes/figures/sticker.png" alt="mistr Logo" width="200"/>

  # mistr

  [![R](https://img.shields.io/badge/R-%23E67E22.svg?&logo=R&logoColor=white)](https://www.r-project.org/)
  [![CRAN Status](https://www.r-pkg.org/badges/version/mistr)](https://cran.r-project.org/package=mistr)
  [![License: GPL-3.0](https://img.shields.io/badge/License-GPL%203.0-blue.svg)](https://opensource.org/licenses/GPL-3.0)

  *A computational framework for mixture and composite distributions.*

  [Key Features](#key-features) •
  [Project Overview](#project-overview) •
  [Citation](#citation)
</div>

---

## Key Features  

📦 **Flexible Distribution Modeling**
- Comprehensive support for standard, composite, and mixture distributions.
- Built-in symbolic differentiation engine for transformations and inverse transformations, simplifying complex operations on random variables.

📊 **Advanced Statistical Functions**
- Seamlessly evaluates PDFs, CDFs, quantiles, and random samples.
- Includes monotonic transformations and mixtures with custom weights.

🧮 **Efficient Mathematical Framework**
- Automatic construction of required equations for transformed random variables.
- Supports truncated and hierarchical distributions for specialized modeling.

📖 **Extensive Documentation**
- Detailed vignettes and examples, including tutorials on risk measures, parameter estimation, and real-world applications.

## Project Overview

`mistr` is a cutting-edge R package that enables users to define, manipulate, and analyze univariate probability distributions. Whether modeling complex financial data or designing custom distributions, `mistr` provides the tools needed for a wide range of statistical tasks.

Highlights include:
- **Composite and Mixture Models**: Easily construct and analyze hierarchical and truncated distributions.
- **Symbolic Differentiation**: Built-in engine to derive and manage equations for transformations and inversions of random variables.
- **Versatility**: From academic research to real-world applications, `mistr` simplifies handling distributions.

### Applications
- **Financial Mathematics**: Model heavy-tailed data for asset returns, risk management, and loss modeling.
- **Actuarial Science**: Fit and evaluate composite models for insurance claim distributions.
- **Custom Distributions**: Create and analyze tailored distributions for experimental data.

## Citation

If you use `mistr` in your research, please cite:

```bibtex
@article{mistr2020,
  title = "{mistr: A Computational Framework for Mixture and Composite Distributions}",
  author = {Lukas Sablica and Kurt Hornik},
  journal = "{The R Journal}",
  year = "2020",
  volume = "12",
  number = "1",
  pages = "283--299",
  doi = "10.32614/RJ-2020-003",
  url = "https://journal.r-project.org/archive/2020/RJ-2020-003/index.html"
}
```





NEWS.md
# mistr 0.0.7

## Minor improvements and fixes

*

# mistr 0.0.6

## Minor improvements and fixes

* tweaked citation

# mistr 0.0.5

## Minor improvements and fixes

* corrected ylab1 argument, now using plotmath  

* removing direct linking to quantmod in help files

# mistr 0.0.4

## Minor improvements and fixes

* corrected citation  

# mistr 0.0.3

## Minor improvements and fixes

* new function `q_approxfun()` that for a given object generates function that approximates the quantile function

* warnings were added to the `risk()` methods if some necessary moments of the distribution do not exist

* citation was added 


# mistr 0.0.2

## Minor improvements and fixes

* the `cols` argument in `autoplot()` has been renamed to `ncols` (fix for the partial match of `col` parameter) 



Work/sablica-hornik/sablica-hornik.R
## This file contains the source code that was used to generate the article.
## The code and the results are discussed in the article.

# Introduction example
## generating SAP density and Q-Q plot
library(mistr)
v = 4
op <- par(mfrow = c(1, 2), mar = c(v,v/2,v/2,2*v))

plot(density(stocks$SAP), xlim = c(-0.07, 0.07), xlab = "", ylab = "", main = "", cex.axis = 0.75)
title(main= "Density of SAP log-returns
     and normal distribution ", cex.main=0.75)
x <- seq(-0.07, 0.07, 0.001)
lines(x, dnorm(x, mean(stocks$SAP), sd(stocks$SAP)), col = "red")
legend("topright", legend = c("SAP", "Normal"), col = c(1, 2), lty = 1, cex = 0.5)
op <- par(mfrow = c(1, 2), mar = c(v,2*v,v/2,v/2))
qqnorm(stocks$SAP, xlab = "", ylab = "", main = "", cex.axis = 0.75)
qqline(stocks$SAP)
title(main= "Normal Q-Q Plot", cex.main=0.75)

par(op)


# Computational framework
N <- normdist(mean = 1, sd = 3) # creates normal distribution with mean = 1 and sd = 3
N                               # prints the distreibution

d(N, c(1, 2, 3))                # PDF evaluation
p(N, c(1, 2, 3))                # CDF evaluation
q(N, c(0.1, 0.2, 0.3))          # QF evaluation
r(N, 3)                         # random numbers generation

B <- binomdist(size = 12, prob = 0.3)   # creates binomial distribution
plim(B, c(-3, 0, 3, 12))                # left-hand limit of PDF evaluation
qlim(B, plim(B, c(0, 3, 7, 12)))        # pseudo-inverse of of plim evaluation

# Adding transformation
E <- expdist(2)                 # creates exponential distribution
E * 2                           # performs transformation
E^2                             # performs transformation

E2 <- E * -2                    # performs transformation
E3 <- E2 * 5                    # performs transformation
E3                              # prints the distreibution

Norm_trafo <- (N - 1)^(1/3)     # performs transformation
Norm_trafo                      # prints the distreibution

Binom_trafo <- -3 * log(B + 4)        # performs transformation
q(Binom_trafo, c(0.05, 0.5, 0.95))    # QF evaluation
plim(Binom_trafo, c(-6, -5, 0))       # left-hand limit of PDF evaluation
sudo_support(Binom_trafo)             # prints the range of the support

par(mai = c(0.4, 0.4, 0.2, 0.2))
plot(Norm_trafo, xlim1 = c(-2.5, 2.5), ylab1 = "", cex.axis = 0.75)    # PDF and CDF plot of the Norm_trafo object

library(ggplot2) # autoplot requires ggplot2
autoplot(Norm_trafo, xlim1 = c(-2.5, 2.5))                             # PDF and CDF plot of the Norm_trafo object using ggplot2

QQplotgg(Norm_trafo, r(Norm_trafo, 1000), conf = 0.99, ylab = NULL, xlab = NULL)   # Q-Qplot of Norm_trafo against its sample

# Mixtures
mixdist(c("norm", "unif"), list(c(2, 2), c(1, 5)), weights = c(0.5, 0.5))        # creates mixture distribution

M <- mixdist(Norm_trafo, Binom_trafo, expdist(0.5), weights = c(0.4, 0.2, 0.4))  # combines objects into mixture distribution

DM <- mixdist(3 * binomdist(12, 0.4), -2*poisdist(2) + 12, weights=c(0.5, 0.5))  # following code was used to test the QF of the mixture distribution
y <- c(0.05, 0.4, p(-DM, c(-5, -10, -15)), 0.95)
x <- q(-DM, y)
autoplot(-DM, which = "cdf", only_mix = TRUE, xlim1 = c(-37, 0)) +
  annotate("point", x, y, col = "white")

sudo_support(M)                       # prints the range of the support

M_trans <- -2 * (M)^(1/3)             # performs transformation
r(M_trans, 4)                         # random numbers generation

autoplot(M_trans)                      # PDF and CDF plot of the M_trans object using ggplot2

# Composite distributions
C <- compdist(-paretodist(1, 1), normdist(0, 2), geomdist(0.3) + 2,           # creates composite distribution
              weights = c(0.15, 0.7, 0.15), breakpoints = c(-3, 3),
              break.spec = c("L", "R"))
C                                                                             # prints the distreibution

C2 <- compdist(-expdist(2), poisdist(), expdist(2),                           # creates composite distribution
               weights = c(0.25, 0.5, 0.25), breakpoints = c(0, 0))
C2                                                                            # prints the distreibution

par(mai = c(0.4, 0.4, 0.2, 0.2))
plot(C, xlim1 = c(-15, 15), ylab1 = "", cex.axis = 0.75, mtext_cex = 0.75)    # PDF and CDF plot of the C object

autoplot(C2, text_ylim = 0.01)                                                # PDF and CDF plot of the C2 object using ggplot2

C_trans <- -0.5 * (C + 7)                                                     # performs transformation

q(C_trans, c(0.075, 0.5, 0.7, 0.9))                                           # QF evaluation
r(C_trans, 4)                                                                 # random numbers generation

autoplot(C_trans, xlim1 = c(-10,5))                                           # PDF and CDF plot of the C_trans object using ggplot2

# Combining mixture and composite distributions
C3 <- compdist(M_trans - 3, C_trans, weights = c(0.5, 0.5), breakpoints = -4.5)        # creates composite distribution
C3_trans <- -2 * C3 + 2                                                                # performs transformation

plim(C3_trans, c(6, 10, 12))                                                           # left-hand limit of PDF and its pseudo-inverse evaluation
qlim(C3_trans, c(0.3, 0.5, 0.7))

autoplot(C3_trans, xlim1 = c(0,20), text_ylim = 0.01, grey = TRUE)                     # PDF and CDF plot of the C3_trans object using ggplot2

autoplot(mixdist( C3_trans, C2 + 5, weights = c(0.7, 0.3)), xlim1 = c(0, 15))          # PDF and CDF plot of the created mixture object using ggplot2

# Data modeling
PNP_model <- PNP_fit(stocks$SAP)                             # estimates PNP model
PNP_model                                                    # prints PNP model

plot(PNP_model, ylab1 = "", ylab2 = "")                      # PDF, CDF and QQ plot of the PNP model

GNG_model <- GNG_fit(stocks$SAP)                             # estimates GNG model
GNG_model                                                    # prints GNG model

autoplot(GNG_model)                                          # PDF, CDF and QQ plot of the GNG model using ggplot2

risk(GNG_model, c(0.02, 0.05, 0.07, 0.1, 0.2, 0.3))          # calculates risk measures of the GNG model for the levels 0.02, 0.05, 0.07, 0.1, 0.2, 0.3


R/compdist.R
#' @title Creates an Object Representing Composite Distribution
#' @description  \code{compdist} creates an object which represents the composite distribution.
#' @param ... distribution objects.
#' @param weights vector of weights for the components.
#' @param breakpoints vector of breakpoints for the composite models, first and last breakpoints (\eqn{-\infty}, \eqn{\infty})
#'                    are assumed to be given, and should not be specified.
#' @param break.spec vector of breakpoints specifications with values "L" or "R", breakpoints specifications corresponding to
#'                   \eqn{-\infty} and \eqn{\infty} should not be specified.
#' @param all.left if TRUE, all break.spec are set to "L", default: FALSE.
#' @param dist vector of distribution names.
#' @param params list of parameters.
#' @return Object of class compdist.
#' @details A CDF of a composite distribution function is \deqn{F(A)=\sum w_{i}F_{i}(A|B_{i})}, where
#'          \eqn{w_{i}} is the weight of the i-th component, \eqn{F_{i}()} is the CDF of the i-th component
#'          and \eqn{B_{i}} is the interval specified by the breakpoints. Clearly, the composite models are a
#'          specific case of the mixture models, where the corresponding probability distribution
#'          functions are truncated to some disjoint support.
#'
#'          The objects can be specified in two ways, either the user may enter
#'          objects representing distributions or a vector of names and list of parameters.
#'          See the examples below.
#'
#'          The argument \code{break.spec} defines if the breakpoint should be included to the distribution
#'          to the right ("R") or to the left ("L") of the breakpoint. This feature is of course useful
#'          only in the case where at least one of the adjacent components is discrete. By default
#'          the intervals are left-closed (all \code{break.spec} values are "R").
#'
#'          The function permits to use the same breakpoint twice. This possibility
#'          allows to define a partition on a singleton, and hence to create a mass of probability.
#'          If this feature is used, the break.spec needs to be specified with "R" and "L", for
#'          the first and the second identical breakpoints, respectively, or not set at all.
#'
#' @examples
#' # using the objects
#' C <- compdist(normdist(1, 3), expdist(4), weights = c(0.7, 0.3), breakpoints = 2)
#' C
#'
#' # using the names and parameters
#' C2 <- compdist(c("norm","exp"), list(c(mean = 1, sd = 3), c(rate = 4)),
#'               weights = c(0.7, 0.3), breakpoints = 2)
#' C2
#'
#' # more complicated model where break.spec is useful
#' C3 <- compdist(-GPDdist(1,0.15,0.7), normdist(-1,1), binomdist(5,0.5),
#'               geomdist(0.3) + 2, weights = c(0.075, 0.425, 0.425, 0.075),
#'               breakpoints = c(-2.5, 0,3), break.spec = c("L", "R", "R"))
#' C3
#'
#' # same breakpoint twice
#' C4 <-  compdist(-expdist(2),poisdist(),expdist(2),
#'                 weights = c(0.25, 0.5, 0.25), breakpoints = c(0, 0))
#' C4
#' @export
#' @rdname compdist
#' @seealso \code{\link{mixdist}}
compdist <- function(..., weights, breakpoints, break.spec, all.left = FALSE) UseMethod("compdist")


findInterval2 <- function(x, vec, spec) {
  if (all(spec == "L"))
    return(findInterval(x, vec, left.open = TRUE))
  if (all(spec == "R"))
    return(findInterval(x, vec))
  z <- numeric(length(x))
  spec <- c(spec, "L")
  l <- findInterval(x, vec)
  r <- findInterval(x, vec, left.open = TRUE)
  if (any(duplicated(vec))) {
    l[(l - r) == 2] <- l[(l - r) == 2] - 1
    r[(l - r) == 2] <- l[(l - r) == 2] - 1
  }
  for (i in unique(c(l[l > 0], r[r > 0]))) {
    if (spec[i] == "R") {
      if (spec[i + 1] == "R") {
        # [)
        z[l == i] <- i
      } else {
        # RL []
        z[r == i | l == i] <- i
      }
    } else {
      if (spec[i + 1] == "R") {
        # ()
        z[r == i & l == i] <- i
      } else {
        # (]
        z[r == i] <- i
      }
    }
  }
  z
}


lambda <- function(obj, br, spc) {
  c(0, unlist(lapply(2:length(obj), function(i) {
    if (spc[i - 1] == "L") p(obj[[i]], br[i - 1]) else plim(obj[[i]], br[i - 1])
  })))
}

rho <- function(obj, br, spc) {
  c(unlist(lapply(1:(length(obj) - 1), function(i) {
    if (spc[i] == "L") p(obj[[i]], br[i]) else plim(obj[[i]], br[i])
  })), 1)
}

#' @rdname compdist
#' @export
compdist.dist <- function(..., weights, breakpoints, break.spec, all.left = FALSE) {
  K <- list(...)
  if (length(K) != length(weights))
    stop("you need the same length of weights and distributions")
  if (length(K) != (length(breakpoints) + 1))
    stop("you need n-1 breakpoints")
  if (any(duplicated(breakpoints[duplicated(breakpoints)])))
    stop("same breakpoint can be used only once or twice")
  if (sum(weights) != 1)
    weights <- weights/sum(weights)
  o <- order(breakpoints)
  breakpoints <- breakpoints[o]
  if (any(duplicated(breakpoints))) {
    d <- duplicated(breakpoints)
    if (!all.left && missing(break.spec)) {
      interval <- rep.int("R", length(breakpoints))
      interval[d] <- "L"
    } else if (all.left && missing(break.spec)) {
      interval <- rep.int("L", length(breakpoints))
      interval[which(d) - 1] <- "R"
    } else {
      break.spec <- break.spec[o]
      if (length(break.spec) != length(breakpoints))
        stop("length of breakpoints and it's specification has to be equal")
      if (any(break.spec != "R" & break.spec != "L"))
        stop("break.spec has to be specified using \"L\" and \"R\"")
      if (any(break.spec[d] == "R" | break.spec[which(d) - 1] == "L"))
        stop("duplicated breakpoints need to be specified as R and L, otherwise there is no support for the distribution")
      interval <- break.spec
    }
  } else {
    if (!all.left && missing(break.spec)){
      interval <- rep.int("R", length(breakpoints))
    } else if (all.left && missing(break.spec)){
      interval <- rep.int("L", length(breakpoints))
    } else {
      break.spec <- break.spec[o]
      if (length(break.spec) != length(breakpoints))
        stop("length of breakpoints and it's specification has to be equal")
      if (any(break.spec != "R" & break.spec != "L"))
        stop("break.spec has to be specified using \"L\" and \"R\"")
      interval <- break.spec
    }
  }
  x <- list(objects = K, weights = weights, breakpoints = breakpoints, interval = interval)
  rhos <- rho(obj = K, br = breakpoints, spc = interval)
  lambdas <- lambda(obj = K, br = breakpoints, spc = interval)
  diff <- rhos - lambdas
  if (any(diff == 0))
    stop("all components need to have a positive support on the truncated interval")
  x <- c(x, list(trunc = list(lambda = lambdas, rho = rhos, diff = diff)))
  if (all(unlist(lapply(K, is.contin)))) {
    class(x) <- c("contcompdist", "compdist", "univdist", "dist")
  } else if (all(unlist(lapply(K, is.discrete)))) {
    class(x) <- c("discrcompdist", "compdist", "univdist", "dist")
  } else {
    class(x) <- c("contdiscrcompdist", "compdist", "univdist", "dist")
  }
  x
}
#' @rdname compdist
#' @export
compdist.default <- function(dist, params, weights, breakpoints, break.spec, all.left = FALSE,...) {
  if (length(dist) != length(weights))
    stop("you need the same length of probabilities and distributions")
  if (length(dist) != (length(breakpoints) + 1))
    stop("you need n-1 breakpoints")
  if (any(duplicated(breakpoints[duplicated(breakpoints)])))
    stop("same breakpoint can be used only once or twice")
  if (sum(weights) != 1)
    weights <- weights/sum(weights)
  K <- lapply(1:length(dist), function(i) {
    D <- get(paste0(dist[i], "dist"), mode = "function")
    do.call(D, as.list(params[[i]]))
  })
  o <- order(breakpoints)
  breakpoints <- breakpoints[o]
  if (any(duplicated(breakpoints))) {
    d <- duplicated(breakpoints)
    if (!all.left && missing(break.spec)) {
      interval <- rep.int("R", length(breakpoints))
      interval[d] <- "L"
    } else if (all.left && missing(break.spec)) {
      interval <- rep.int("L", length(breakpoints))
      interval[which(d) - 1] <- "R"
    } else {
      break.spec <- break.spec[o]
      if (length(break.spec) != length(breakpoints))
        stop("length of breakpoints and it's specification has to be equal")
      if (any(break.spec != "R" & break.spec != "L"))
        stop("break.spec has to be specified using \"L\" and \"R\"")
      if (any(break.spec[d] == "R" | break.spec[which(d) - 1] == "L"))
        stop("duplicated breakpoints need to be specified as R and L, otherwise there is no support for the distribution")
      interval <- break.spec
    }
  } else {
    if (!all.left && missing(break.spec))
      interval <- rep.int("R", length(breakpoints)) else if (all.left && missing(break.spec))
        interval <- rep.int("L", length(breakpoints)) else {
          break.spec <- break.spec[o]
          if (length(break.spec) != length(breakpoints))
            stop("length of breakpoints and it's specification has to be equal")
          if (any(break.spec != "R" & break.spec != "L"))
            stop("break.spec has to be specified using \"L\" and \"R\"")
          interval <- break.spec
        }
  }
  x <- list(objects = K, weights = weights, breakpoints = breakpoints, interval = interval)
  rhos <- rho(obj = K, br = breakpoints, spc = interval)
  lambdas <- lambda(obj = K, br = breakpoints, spc = interval)
  diff <- rhos - lambdas
  if (any(diff == 0))
    stop("all components need to have a positive support on the truncated interval")
  x <- c(x, list(trunc = list(lambda = lambdas, rho = rhos, diff = diff)))
  
  if (all(unlist(lapply(K, function(O) {
    any(class(O) == "contdist")
  })))) {
    class(x) <- c("contcompdist", "compdist", "univdist", "dist")
  } else if (all(unlist(lapply(K, function(O) {
    any(class(O) == "discrdist")
  })))) {
    class(x) <- c("discrcompdist", "compdist", "univdist", "dist")
  } else {
    class(x) <- c("contdiscrcompdist", "compdist", "univdist", "dist")
  }
  x
}

#' @export
print.compdist <- function(x, digits = 4, ...) {
  Distribution <- numeric(length(x$objects))
  Parameters <- numeric(length(x$objects))
  mix <- unlist(lapply(x$objects, is.mixture), use.names = FALSE)
  comp <- unlist(lapply(x$objects, is.composite), use.names = FALSE)
  dist <- !(mix | comp)
  Distribution[mix] <- "Mixture distribution"
  Distribution[comp] <- "Composite distribution"
  Distribution[dist] <- unlist(lapply(x$objects[dist], function(x) x$type), use.names = FALSE)
  Trafo <- sapply(x$objects, function(x) {
    if (is.transformed(x)) {
      deparse(x$trafo$print)
    } else "none"
  })
  Parameters[dist] <- unlist(lapply(x$objects[dist], function(x) {
    paste(names(x$parameters), round(unlist(x$parameters), digits), sep = " = ", collapse = ", ")
  }), use.names = FALSE)
  Parameters[!dist] <- "   "
  Weight <- round(x$weights, digits)
  vv <- unlist(lapply(x$interval, function(x) if (x == "R")
    c(")", "[") else c("]", "(")))
  bre <- round(x$breakpoints, digits)
  Truncation <- paste(c("(", vv[seq.int(1L, length(vv), 2L) + 1]), c("-Inf", bre), ",", c(bre, "Inf"), c(vv[seq.int(1L,
                                                                                                                    length(vv), 2L)], ")"), sep = "")
  if (any(Trafo != "none")) {
    results <- data.frame(Trafo, Distribution, Parameters, Weight, Truncation)
  } else {
    results <- data.frame(Distribution, Parameters, Weight, Truncation)
  }
  name.width <- unlist(lapply(names(results), nchar), use.names = FALSE)
  cat("Composite distribution with: \n \n")
  print(format(results, width = name.width, justify = "centre"))
  invisible(x)
}

#' @rdname d
#' @export
d.compdist <- function(O, x, log = FALSE) {
  Z <- numeric(length(x))
  nat <- is.na(x)
  Z[nat] <- x[nat]
  x <- x[!nat]
  z <- numeric(length(x))
  int <- findInterval2(x, O$breakpoints, O$interval) + 1
  un <- sort(unique(int))
  dif <- O$trunc$diff[un]
  if (log) {
    for (j in seq_along(un)) {
      i <- un[j]
      z[int == i] <- log(O$weights[i]) + d(O$objects[[i]], x[int == i], log = TRUE) - log(dif[j])
    }
  } else {
    for (j in seq_along(un)) {
      i <- un[j]
      z[int == i] <- O$weights[i] * d(O$objects[[i]], x[int == i])/dif[j]
    }
  }
  Z[!nat] <- z
  Z
}

#' @rdname p
#' @export
p.compdist <- function(O, q, lower.tail = TRUE, log.p = FALSE) {
  Z <- numeric(length(q))
  nat <- is.na(q)
  Z[nat] <- q[nat]
  q <- q[!nat]
  z <- numeric(length(q))
  w <- cumsum(c(0, O$weights))
  int <- findInterval2(q, O$breakpoints, O$interval) + 1
  un <- sort(unique(int))
  lam <- O$trunc$lambda[un]
  dif <- O$trunc$diff[un]
  for (j in seq_along(un)) {
    i <- un[j]
    z[int == i] <- w[i] + O$weights[i] * (p(O$objects[[i]], q[int == i]) - lam[j])/dif[j]
  }
  Z[!nat] <- z
  if (!lower.tail)
    Z <- 1 - Z
  if (log.p)
    log(Z) else Z
}
#' @rdname plim
#' @export
plim.compdist <- function(O, q, lower.tail = TRUE, log.p = FALSE) {
  w <- cumsum(c(0, O$weights))
  Z <- numeric(length(q))
  nat <- is.na(q)
  Z[nat] <- q[nat]
  q <- q[!nat]
  z <- numeric(length(q))
  int <- findInterval(q, O$breakpoints, left.open = TRUE) + 1
  un <- sort(unique(int))
  lam <- O$trunc$lambda[un]
  dif <- O$trunc$diff[un]
  for (j in seq_along(un)) {
    i <- un[j]
    z[int == i] <- w[i] + O$weights[i] * (plim(O$objects[[i]], q[int == i]) - lam[j])/dif[j]
  }
  Z[!nat] <- z
  if (!lower.tail)
    Z <- 1 - Z
  if (log.p)
    log(Z) else Z
}
#' @rdname q
#' @export
q.compdist <- function(O, p, lower.tail = TRUE, log.p = FALSE, ...) {
  Z <- numeric(length(p))
  nat <- is.na(p)
  Z[nat] <- p[nat]
  p <- p[!nat]
  z <- numeric(length(p))
  if (log.p)
    p <- exp(p)
  if (!lower.tail)
    p <- 1 - p
  ok <- p >= 0 & p <= 1
  p <- p[ok]
  zz <- numeric(length(p))
  zz[p == 1] <- unname(sudo_support(O)[2])
  zz[p == 0] <- unname(sudo_support(O)[1])
  supp <- p > 0 & p < 1
  p <- p[supp]
  zzz <- numeric(length(p))
  w <- cumsum(c(0, O$weights))
  int <- findInterval(p, w, left.open = TRUE)
  un <- sort(unique(int))
  lam <- O$trunc$lambda[un]
  dif <- O$trunc$diff[un]
  for (j in seq_along(un)) {
    i <- un[j]
    zzz[int == i] <- q(O$objects[[i]], lam[j] + ((p[int == i] - w[i]) * dif[j])/O$weights[i])
  }
  zz[supp] <- zzz
  z[ok] <- zz
  z[!ok] <- NaN
  Z[!nat] <- z
  Z
}
#' @rdname qlim
#' @export
qlim.compdist <- function(O, p, lower.tail = TRUE, log.p = FALSE) {
  Z <- numeric(length(p))
  nat <- is.na(p)
  Z[nat] <- p[nat]
  p <- p[!nat]
  z <- numeric(length(p))
  if (log.p)
    p <- exp(p)
  if (!lower.tail)
    p <- 1 - p
  ok <- p >= 0 & p <= 1
  p <- p[ok]
  zz <- numeric(length(p))
  zz[p == 1] <- unname(sudo_support(O)[2])
  zz[p == 0] <- unname(sudo_support(O)[1])
  supp <- p > 0 & p < 1
  p <- p[supp]
  zzz <- numeric(length(p))
  w <- cumsum(c(0, O$weights))
  int <- findInterval(p, w)
  un <- sort(unique(int))
  lam <- O$trunc$lambda[un]
  dif <- O$trunc$diff[un]
  for (j in seq_along(un)) {
    i <- un[j]
    zzz[int == i] <- qlim(O$objects[[i]], lam[j] + ((p[int == i] - w[i]) * dif[j])/O$weights[i])
  }
  zz[supp] <- zzz
  z[ok] <- zz
  z[!ok] <- NaN
  Z[!nat] <- z
  Z
}
#' @rdname r
#' @export
r.compdist <- function(O, n) {
  if (any(sapply(O$objects, is.mixture)))
    r2(O, n) else q(O, runif(n))
}

r2 <- function(O, n) {
  Z <- numeric(n)
  l <- length(O$objects)
  t <- sample(l, replace = TRUE, n, prob = O$weights)
  v <- unlist(lapply(1:l, function(x) sum(t == x)), use.names = FALSE)
  lam <- O$trunc$lambda
  dif <- O$trunc$diff
  lapply(1:l, function(i) {
    if (is.mixture(O$objects[[i]])) {
      h <- numeric(v[i])
      n <- 0
      num <- min(round(2 * v[i]/dif[i]), 1e+07)
      while (n < v[i]) {
        o <- r(O$objects[[i]], num)
        int <- findInterval2(o, O$breakpoints, O$interval) + 1
        good <- o[int == i]
        tt <- length(good)
        if (tt == 0)
          next
        if (tt + n > v[i]) {
          h[(n + 1):v[i]] <- good[1:(v[i] - n)]
          n <- v[i]
        } else {
          h[(n + 1):(n + tt)] <- good
          n <- n + tt
        }
      }
      Z[t == i] <<- h
    } else {
      Z[t == i] <<- q(O$objects[[i]], lam[i] + runif(v[i]) * dif[i])
    }
  })
  Z
}

#' @export
print.trans_compdist <- function(x, digits = 4, ...) {
  Distribution <- numeric(length(x$objects))
  Parameters <- numeric(length(x$objects))
  mix <- unlist(lapply(x$objects, is.mixture), use.names = FALSE)
  comp <- unlist(lapply(x$objects, is.composite), use.names = FALSE)
  dist <- !(mix | comp)
  Distribution[mix] <- "Mixture distribution"
  Distribution[comp] <- "Composite distribution"
  Distribution[dist] <- unlist(lapply(x$objects[dist], function(x) x$type), use.names = FALSE)
  Trafo <- sapply(x$objects, function(x) {
    if (is.transformed(x)) {
      deparse(x$trafo$print)
    } else "none"
  })
  Parameters[dist] <- unlist(lapply(x$objects[dist], function(x) {
    paste(names(x$parameters), round(unlist(x$parameters), digits), sep = " = ", collapse = ", ")
  }), use.names = FALSE)
  Parameters[!dist] <- "   "
  Weight <- round(x$weights, digits)
  vv <- unlist(lapply(x$interval, function(x) if (x == "R")
    c(")", "[") else c("]", "(")))
  bre <- round(x$breakpoints, digits)
  Truncation <- paste(c("(", vv[seq.int(1L, length(vv), 2L) + 1]), c("-Inf", bre), ",", c(bre, "Inf"), c(vv[seq.int(1L,
                                                                                                                    length(vv), 2L)], ")"), sep = "")
  if (any(Trafo != "none")) {
    results <- data.frame(Trafo, Distribution, Parameters, Weight, Truncation)
  } else {
    results <- data.frame(Distribution, Parameters, Weight, Truncation)
  }
  name.width <- unlist(lapply(names(results), nchar), use.names = FALSE)
  cat("Monotonically transformed composite distribution with: \n \n")
  cat("Trafo: ", deparse(x$trafo$print), "\n \n")
  print(format(results, width = name.width, justify = "centre"))
  invisible(x)
}

change_support <- function(O, sup) UseMethod("change_support")
#' @export
change_support.standist <- function(O, sup) {
  sup <- sort(sup)
  if (O$support$to < sup[1] || O$support$from > sup[2]) {
    O$support$from <- sup[1]
    O$support$to <- sup[2]
  } else {
    O$support$from <- max(sup[1], O$support$from)
    O$support$to <- min(sup[2], O$support$to)
  }
  O
}
#' @export
change_support.trans_standist <- function(O, sup) {
  sup <- sort(eval(O$trafo$invtrans, list(X = sup)))
  if (O$support$to < sup[1] || O$support$from > sup[2]) {
    O$support$from <- sup[1]
    O$support$to <- sup[2]
  } else {
    O$support$from <- max(sup[1], O$support$from)
    O$support$to <- min(sup[2], O$support$to)
  }
  O
}
#' @export
change_support.mixdist <- function(O, sup) {
  O$objects <- lapply(O$objects, change_support, sup)
  O
}
#' @export
change_support.trans_mixdist <- function(O, sup) {
  sup <- sort(eval(O$trafo$invtrans, list(X = sup)))
  O$objects <- lapply(O$objects, change_support, sup)
  O
}
#' @export
change_support.compdist <- function(O, sup) {
  O$objects <- lapply(O$objects, change_support, sup)
  O
}
#' @export
change_support.trans_compdist <- function(O, sup) {
  sup <- sort(eval(O$trafo$invtrans, list(X = sup)))
  O$objects <- lapply(O$objects, change_support, sup)
  O
}


#' @rdname d
#' @export
d.trans_compdist <- function(O, x, log = FALSE) {
  Z <- numeric(length(x))
  nat <- is.na(x)
  Z[nat] <- x[nat]
  x <- x[!nat]
  zz <- numeric(length(x))
  r <- x >= sudo_support(O)[1] & x <= sudo_support(O)[2]
  x <- x[r]
  z <- numeric(length(x))
  sup <- eval(O$trafo$invtrans, list(X = sudo_support(O)))
  if (monot(O) == 1) {
    spec <- O$interval
    obj <- lapply(O$objects, function(x) eval(O$trafo$print, list(X = change_support(x, sup))))
    weigh <- O$weights
    dif <- O$trunc$diff
  } else {
    spec <- sapply(rev(O$interval), function(x) if (x == "L")
      "R" else "L")
    obj <- lapply(rev(O$objects), function(x) eval(O$trafo$print, list(X = change_support(x, sup))))
    dif <- rev(O$trunc$diff)
    weigh <- rev(O$weights)
  }
  breakp <- sort(round(eval(O$trafo$trans, list(X = O$breakpoints)), 14))
  int <- findInterval2(x, breakp, spec) + 1
  un <- sort(unique(int))
  dif <- dif[un]
  if (log) {
    for (j in seq_along(un)) {
      i <- un[j]
      z[int == i] <- log(weigh[i]) + d(obj[[i]], x[int == i], log = TRUE) - log(dif[j])
    }
    zz[r == 1] <- z
    zz[r != 1] <- -Inf
  } else {
    for (j in seq_along(un)) {
      i <- un[j]
      z[int == i] <- weigh[i] * d(obj[[i]], x[int == i])/dif[j]
    }
    zz[r == 1] <- z
  }
  Z[!nat] <- zz
  Z
}
